File: codegen_test1/cpp/shared.cpp
#include "shared.h"
#include "shared.cpp.inc"
void CodeGen_shared::Register() {
	::xx::ObjManager::Register<::A>();
	::xx::ObjManager::Register<::B>();
}
namespace xx {
	void ObjFuncs<::C, void>::Write(::xx::ObjManager& om, ::xx::Data& d, ::C const& in) {
        om.Write(d, in.x);
        om.Write(d, in.y);
        om.Write(d, in.targets);
    }
	void ObjFuncs<::C, void>::WriteFast(::xx::ObjManager& om, ::xx::Data& d, ::C const& in) {
        om.Write<false>(d, in.x);
        om.Write<false>(d, in.y);
        om.Write<false>(d, in.targets);
    }
	int ObjFuncs<::C, void>::Read(::xx::ObjManager& om, ::xx::Data& d, ::C& out) {
        if (int r = om.Read(d, out.x)) return r;
        if (int r = om.Read(d, out.y)) return r;
        if (int r = om.Read(d, out.targets)) return r;
        return 0;
    }
	void ObjFuncs<::C, void>::Append(ObjManager &om, std::string& s, ::C const& in) {
#ifndef XX_DISABLE_APPEND
        s.push_back('{');
        AppendCore(om, s, in);
        s.push_back('}');
#endif
    }
	void ObjFuncs<::C, void>::AppendCore(ObjManager &om, std::string& s, ::C const& in) {
#ifndef XX_DISABLE_APPEND
        om.Append(s, "\"x\":", in.x); 
        om.Append(s, ",\"y\":", in.y);
        om.Append(s, ",\"targets\":", in.targets);
#endif
    }
    void ObjFuncs<::C>::Clone(::xx::ObjManager& om, ::C const& in, ::C &out) {
        om.Clone_(in.x, out.x);
        om.Clone_(in.y, out.y);
        om.Clone_(in.targets, out.targets);
    }
    int ObjFuncs<::C>::RecursiveCheck(::xx::ObjManager& om, ::C const& in) {
        if (int r = om.RecursiveCheck(in.x)) return r;
        if (int r = om.RecursiveCheck(in.y)) return r;
        if (int r = om.RecursiveCheck(in.targets)) return r;
        return 0;
    }
    void ObjFuncs<::C>::RecursiveReset(::xx::ObjManager& om, ::C& in) {
        om.RecursiveReset(in.x);
        om.RecursiveReset(in.y);
        om.RecursiveReset(in.targets);
    }
    void ObjFuncs<::C>::SetDefaultValue(::xx::ObjManager& om, ::C& in) {
        in.x = 0.0f;
        in.y = 0.0f;
        om.SetDefaultValue(in.targets);
    }
}
void A::Write(::xx::ObjManager& om, ::xx::Data& d) const {
    auto bak = d.WriteJump(sizeof(uint32_t));
    om.Write(d, this->id);
    om.Write(d, this->nick);
    om.Write(d, this->parent);
    om.Write(d, this->children);
    d.WriteFixedAt(bak, (uint32_t)(d.len - bak));
}
int A::Read(::xx::ObjManager& om, ::xx::Data& d) {
    uint32_t siz;
    if (int r = d.ReadFixed(siz)) return r;
    auto endOffset = d.offset - sizeof(siz) + siz;

    if (d.offset >= endOffset) this->id = 0;
    else if (int r = om.Read(d, this->id)) return r;
    if (d.offset >= endOffset) om.SetDefaultValue(this->nick);
    else if (int r = om.Read(d, this->nick)) return r;
    if (d.offset >= endOffset) om.SetDefaultValue(this->parent);
    else if (int r = om.Read(d, this->parent)) return r;
    if (d.offset >= endOffset) om.SetDefaultValue(this->children);
    else if (int r = om.Read(d, this->children)) return r;

    if (d.offset > endOffset) return __LINE__;
    else d.offset = endOffset;
    return 0;
}
void A::Append(::xx::ObjManager& om, std::string& s) const {
#ifndef XX_DISABLE_APPEND
    ::xx::Append(s, "{\"__typeId__\":1");
    this->AppendCore(om, s);
    s.push_back('}');
#endif
}
void A::AppendCore(::xx::ObjManager& om, std::string& s) const {
#ifndef XX_DISABLE_APPEND
    ::xx::Append(s, ",\"id\":", this->id);
    ::xx::Append(s, ",\"nick\":", this->nick);
    ::xx::Append(s, ",\"parent\":", this->parent);
    ::xx::Append(s, ",\"children\":", this->children);
#endif
}
void A::Clone(::xx::ObjManager& om, void* const &tar) const {
    auto out = (::A*)tar;
    om.Clone_(this->id, out->id);
    om.Clone_(this->nick, out->nick);
    om.Clone_(this->parent, out->parent);
    om.Clone_(this->children, out->children);
}
int A::RecursiveCheck(::xx::ObjManager& om) const {
    if (int r = om.RecursiveCheck(this->id)) return r;
    if (int r = om.RecursiveCheck(this->nick)) return r;
    if (int r = om.RecursiveCheck(this->parent)) return r;
    if (int r = om.RecursiveCheck(this->children)) return r;
    return 0;
}
void A::RecursiveReset(::xx::ObjManager& om) {
    om.RecursiveReset(this->id);
    om.RecursiveReset(this->nick);
    om.RecursiveReset(this->parent);
    om.RecursiveReset(this->children);
}
void A::SetDefaultValue(::xx::ObjManager& om) {
    this->id = 0;
    om.SetDefaultValue(this->nick);
    om.SetDefaultValue(this->parent);
    om.SetDefaultValue(this->children);
}
void B::Write(::xx::ObjManager& om, ::xx::Data& d) const {
    this->BaseType::Write(om, d);
    om.Write(d, this->data);
    om.Write(d, this->c);
    om.Write(d, this->c2);
    om.Write(d, this->c3);
}
int B::Read(::xx::ObjManager& om, ::xx::Data& d) {
    if (int r = this->BaseType::Read(om, d)) return r;
    if (int r = om.Read(d, this->data)) return r;
    if (int r = om.Read(d, this->c)) return r;
    if (int r = om.Read(d, this->c2)) return r;
    if (int r = om.Read(d, this->c3)) return r;
    return 0;
}
void B::Append(::xx::ObjManager& om, std::string& s) const {
#ifndef XX_DISABLE_APPEND
    ::xx::Append(s, "{\"__typeId__\":2");
    this->AppendCore(om, s);
    s.push_back('}');
#endif
}
void B::AppendCore(::xx::ObjManager& om, std::string& s) const {
#ifndef XX_DISABLE_APPEND
    this->BaseType::AppendCore(om, s);
    ::xx::Append(s, ",\"data\":", this->data);
    ::xx::Append(s, ",\"c\":", this->c);
    ::xx::Append(s, ",\"c2\":", this->c2);
    ::xx::Append(s, ",\"c3\":", this->c3);
#endif
}
void B::Clone(::xx::ObjManager& om, void* const &tar) const {
    this->BaseType::Clone(om, tar);
    auto out = (::B*)tar;
    om.Clone_(this->data, out->data);
    om.Clone_(this->c, out->c);
    om.Clone_(this->c2, out->c2);
    om.Clone_(this->c3, out->c3);
}
int B::RecursiveCheck(::xx::ObjManager& om) const {
    if (int r = this->BaseType::RecursiveCheck(om)) return r;
    if (int r = om.RecursiveCheck(this->data)) return r;
    if (int r = om.RecursiveCheck(this->c)) return r;
    if (int r = om.RecursiveCheck(this->c2)) return r;
    if (int r = om.RecursiveCheck(this->c3)) return r;
    return 0;
}
void B::RecursiveReset(::xx::ObjManager& om) {
    this->BaseType::RecursiveReset(om);
    om.RecursiveReset(this->data);
    om.RecursiveReset(this->c);
    om.RecursiveReset(this->c2);
    om.RecursiveReset(this->c3);
}
void B::SetDefaultValue(::xx::ObjManager& om) {
    this->BaseType::SetDefaultValue(om);
    om.SetDefaultValue(this->data);
    om.SetDefaultValue(this->c);
    om.SetDefaultValue(this->c2);
    om.SetDefaultValue(this->c3);
}

File: codegen_test1/lua/shared.lua
require "objmgr"

CodeGen_shared = {
    md5 = "#*MD5<115d3d5cd445941f39210ad7c08d16ff>*#",
    Register = function()
        local o = ObjMgr
        o.Register(A)
        o.Register(B)
    end
}

A = {
    typeName = "A",
    typeId = 1,
    Create = function(c)
        if c == nil then
            c = {}
            setmetatable(c, A)
        end
        c.id = 0 -- Int32
        c.nick = null -- Nullable<String>
        c.parent = null -- Weak<A>
        c.children = {} -- List<Shared<A>>
        return o
    end,
    Read = function(self, om)
        local d = om.d, r, o, len, e
        -- compatible handle
        r, len = d:Ru32()
        if r ~= 0 then return r end
        e = d:GetOffset() - 4 + len
        -- id
        if d:GetOffset() >= e then
            self.id = 0
        else
            r, self.id = d:Ri32()
            if r ~= 0 then return r end
        end
        -- nick
        if d:GetOffset() >= e then
            self.nick = null
        else
            r, self.nick = om:Rnstr()
            if r ~= 0 then return r end
        end
        -- parent
        if d:GetOffset() >= e then
            self.parent = null
        else
            r, self.parent = om:Read()
            if r ~= 0 then return r end
        end
        -- children
        if d:GetOffset() >= e then
            self.children = {}
        else
            r, len = d:Rvu()
            if len > d:GetLeft() then return -1 end
            o = {}
            self.children = o
            for i = 1, len do
                r, o[i] = om:Read()
                if r ~= 0 then return r end
            end
            if r ~= 0 then return r end
        end
        -- compatible handle
        if d:GetOffset() > e then return -1 end
        d:SetOffset(e)
    end,
    Write = function(self, om)
        local d = om.d, o, len
        -- compatible handle
        local bak = d:Wj(4)
        -- id
        d:Wi32(self.id)
        -- nick
        d:Wnstr(self.nick)
        -- parent
        om:Write(self.parent)
        -- children
        o = self.children
        len = #o
        d:Wvu(len)
        for i = 1, len do
            om:Write(o[i])
        end
        -- compatible handle
        d:Wu32_at(bak, d:GetLen() - bak);
    end
}
A.__index = A

B = {
    typeName = "B", -- : A
    typeId = 2,
    Create = function(c)
        if c == nil then
            c = {}
            setmetatable(c, B)
        end
        A.Create(c)
        c.data = NewXxData() -- XxData
        c.c = C.Create() -- C
        c.c2 = null -- Nullable<C>
        c.c3 = {} -- List<Nullable<C>>
        return o
    end,
    Read = function(self, om)
        local d = om.d, r, o
        -- base read
        r = A.Read(self, om)
        if r ~= 0 then return r end
        -- data
        r, self.data = om:Rdata()
        if r ~= 0 then return r end
        -- c
        self.c = C.Create(); r = self.c:Read(om)
        if r ~= 0 then return r end
        -- c2
        r, o = d:Ru8(); if r ~= 0 then return r end; if o == 0 then self.c2 = null else self.c2 = C.Create(); r = self.c2:Read(om) end
        if r ~= 0 then return r end
        -- c3
        r, len = d:Rvu()
        if len > d:GetLeft() then return -1 end
        o = {}
        self.c3 = o
        for i = 1, len do
            r, o = d:Ru8(); if r ~= 0 then return r end; if o == 0 then o[i] = null else o[i] = C.Create(); r = o[i]:Read(om) end
            if r ~= 0 then return r end
        end
        if r ~= 0 then return r end
    end,
    Write = function(self, om)
        local d = om.d, o, len
        -- base read
        A.Write(self, om)
        -- data
        d:Wdata(self.data)
        -- c
        self.c:Write(om)
        -- c2
        if self.c2 == null then d:Wu8(0) else d:Wu8(1); self.c2:Write(om) end
        -- c3
        o = self.c3
        len = #o
        d:Wvu(len)
        for i = 1, len do
            if o[i] == null then d:Wu8(0) else d:Wu8(1); o[i]:Write(om) end
        end
    end
}
B.__index = B

--[[
asdfasdf
]]
C = {
    typeName = "C",
    Create = function(c)
        if c == nil then
            c = {}
            setmetatable(c, C)
        end
        c.x = 0 -- Single
        c.y = 0 -- Single
        c.targets = {} -- List<Weak<A>>
        return o
    end,
    Read = function(self, om)
        local d = om.d, r, o
        -- x
        r, self.x = d:Rf()
        if r ~= 0 then return r end
        -- y
        r, self.y = d:Rf()
        if r ~= 0 then return r end
        -- targets
        r, len = d:Rvu()
        if len > d:GetLeft() then return -1 end
        o = {}
        self.targets = o
        for i = 1, len do
            r, o[i] = om:Read()
            if r ~= 0 then return r end
        end
        if r ~= 0 then return r end
    end,
    Write = function(self, om)
        local d = om.d, o, len
        -- x
        d:Wf(self.x)
        -- y
        d:Wf(self.y)
        -- targets
        o = self.targets
        len = #o
        d:Wvu(len)
        for i = 1, len do
            om:Write(o[i])
        end
    end
}
C.__index = C
